ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validatorcache/ActiveValidatorCache.java

```diff
@@ -0,0 +1,126 @@
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.datastructures.attestation.ValidateableAttestation;
+import tech.pegasys.teku.spec.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.spec.datastructures.operations.IndexedAttestation;
+
+public class ActiveValidatorCache implements ActiveValidatorChannel {
+  private final Spec spec;
+  private static final Logger LOG = LogManager.getLogger();
+  private static final int VALIDATOR_CACHE_SPARE_CAPACITY = 1000;
+  // validatorActiveEpochs is indexed by validator id, contains mod(3) of epoch values - basically
+  // rotating list
+  private UInt64[][] validatorActiveEpochs;
+
+  public ActiveValidatorCache(final Spec spec, final int initialSize) {
+    Preconditions.checkArgument(initialSize >= 0);
+    validatorActiveEpochs = new UInt64[initialSize + VALIDATOR_CACHE_SPARE_CAPACITY][];
+    this.spec = spec;
+  }
+
+  void touch(final UInt64 validatorIndex, final UInt64 epoch) {
+    Preconditions.checkNotNull(validatorIndex);
+    Preconditions.checkNotNull(epoch);
+    LOG.trace("Touch validator {} at epoch {}", validatorIndex, epoch);
+
+    if (validatorIndex.intValue() >= validatorActiveEpochs.length) {
+      grow(validatorIndex);
+    }
+    ensureNodeIsInitialized(validatorIndex);
+    validatorActiveEpochs[validatorIndex.intValue()][epoch.mod(3).intValue()] = epoch;
+  }
+
+  boolean isValidatorSeenAtEpoch(final UInt64 validatorIndex, final UInt64 epoch) {
+    Preconditions.checkNotNull(validatorIndex);
+    Preconditions.checkNotNull(epoch);
+
+    if (validatorActiveEpochs.length <= validatorIndex.intValue()) {
+      LOG.debug(
+          "validator index {} exceeds array size {}", validatorIndex, validatorActiveEpochs.length);
+      return false;
+    }
+
+    return validatorActiveEpochs[validatorIndex.intValue()] != null
+        && epoch.equals(validatorActiveEpochs[validatorIndex.intValue()][epoch.mod(3).intValue()]);
+  }
+
+  int getCacheSize() {
+    return validatorActiveEpochs.length;
+  }
+
+  @VisibleForTesting
+  UInt64[] getValidatorEpochs(final UInt64 validatorIndex) {
+    Preconditions.checkArgument(validatorIndex.intValue() < validatorActiveEpochs.length);
+    return validatorActiveEpochs[validatorIndex.intValue()];
+  }
+
+  private void ensureNodeIsInitialized(final UInt64 validatorIndex) {
+    if (validatorActiveEpochs[validatorIndex.intValue()] == null) {
+      validatorActiveEpochs[validatorIndex.intValue()] = new UInt64[3];
+    }
+  }
+
+  private void grow(final UInt64 requestedValidatorIndex) {
+    final int newSize =
+        requestedValidatorIndex.intValue()
+                < validatorActiveEpochs.length + VALIDATOR_CACHE_SPARE_CAPACITY
+            ? validatorActiveEpochs.length + VALIDATOR_CACHE_SPARE_CAPACITY
+            : requestedValidatorIndex.plus(VALIDATOR_CACHE_SPARE_CAPACITY).intValue();
```

Comment:
I think this is just:
```suggestion
    final int newSize = requestedValidatorIndex.intValue() + VALIDATOR_CACHE_SPARE_CAPACITY
        requestedValidatorIndex.intValue();
```

because if we have to resize the array we want to leave 1000 empty slots at the end. That way we should avoid resizing the array again for another day (since only 900 new validators can join per day).